### PR TITLE
Service selection with equality-based filtering

### PR DESCRIPTION
**Pull request by Palakis**
Solves #1 

Set-based filtering is not yet supported.

TODO:
- [ ] README update

### CODE REVIEW

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java [hunk 1]
```diff
@@ -49,33 +44,80 @@ public String description()
...
+	@Override
+	public List<ServiceInstance> selectInstances(Map<String, String> match, Map<String, String> doNotMatch)
+	{
+		ServiceList serviceList;
+		try {
+			serviceList = kubeClient.services().withLabels(match).withoutLabels(doNotMatch).list();
+		} catch(KubernetesClientException e) {
+			log.warn("getInstancesMatchingMetadata: failed to retrieve matching services: API call failed: {}",
+				e.getMessage());
+			return Collections.emptyList();
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: It's question whether to swallow the exception here - what was your reasoning?

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java [hunk 2]
```diff
@@ -49,33 +44,80 @@ public String description()
...
 		}
 
-		if (service.getSpec().getPorts().isEmpty()) {
-			log.error("getInstances: service '{}' has no ports", serviceId);
+		if (serviceList == null) {
+			log.warn("getInstancesMatchingMetadata: service list is null");
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: This on the other hand seems like perfectly normal case and logging warn seems unnecessary.

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java [hunk 3]
```diff
@@ -49,33 +44,80 @@ public String description()
...
-		List<ServicePort> servicePorts = service.getSpec().getPorts();
-		ServicePort svcPort = servicePorts.get(0); // By default, use the first port available
+		return getInstancesFromServiceList(serviceList.getItems());
+	}
+
+	private List<ServiceInstance> getInstancesFromService(Service service)
+	{
+		List<Service> serviceList = new ArrayList<>();
+		serviceList.add(service);
+		return getInstancesFromServiceList(serviceList);
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: There is a better way of passing single item in a list, see https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#singletonList(T)

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java [hunk 4]
```diff
@@ -49,33 +44,80 @@ public String description()
...
 
-		// But if a port named "http" is available, use it instead
-		Optional<ServicePort> httpPort = servicePorts.stream()
-			.filter(s -> s.getName() != null && s.getName().equals("http")).findFirst();
-		if (httpPort.isPresent()) {
-			svcPort = httpPort.get();
+	private List<ServiceInstance> getInstancesFromServiceList(List<Service> services)
+	{
+		if (log.isDebugEnabled()) {
+			log.debug("services = {}", services.toString());
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: More of a context in the debugging message might be handy, after we forget about this log line.

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java [hunk 5]
```diff
@@ -49,33 +44,80 @@ public String description()
...
+			if (service.getSpec().getPorts().isEmpty()) {
+				log.error("service '{}' has no ports", serviceName);
+				continue;
+			}
+
+			List<ServicePort> servicePorts = service.getSpec().getPorts();
+			ServicePort svcPort = servicePorts.get(0); // By default, use the first port available
+
+			// But if a port named "http" is available, use it instead
+			Optional<ServicePort> httpPort = servicePorts.stream()
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: I find this whole construct a little bit clumsy, I would expect steps like this:

1. if service has only one port, use that one no matter what the name is
2. if service has multiple ports, look for the one with http name
3. if port with http name is found, use that one
4. if port with http name is not found, issue warning to logs (not sure about the level here, either INFO or WARN ...) and use the first one

File: src/main/java/com/dtforce/spring/kubernetes/SelectorEnabledDiscoveryClient.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.cloud.client.discovery.DiscoveryClient;
+
+import java.util.List;
+import java.util.Map;
+
+public interface SelectorEnabledDiscoveryClient extends DiscoveryClient
+{
+
+	// Equality-based matching
+	List<ServiceInstance> selectInstances(Map<String, String> match);
```
Comment (maresja1) [inline src/main/java/com/dtforce/spring/kubernetes/SelectorEnabledDiscoveryClient.java]: As this is an interface, I would appreciate javadoc comments of both methods, including description of parameters.

**Review by maresja1 — CHANGES_REQUESTED** [review]
Thank you @Palakis, please look at review suggestions.

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java (hunk 2 above)
Comment (Palakis) [reply src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: A typical empty query response from the Kubernetes API (and in turn, the client) is an empty list, not a null value. I'll make this clear in the code.

File: src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java (hunk 1 above)
Comment (Palakis) [reply src/main/java/com/dtforce/spring/kubernetes/KubernetesDiscoveryClient.java]: I have to admit: swallowing the exception is not a good idea, as `KubernetesClientException` being thrown means that something is wrong with the API client or permissions of the underlying Service Account.

**Review by maresja1 — APPROVED** [review]

**Pull request merged**